### PR TITLE
[fix] allow saving python attr on Tensor and Parameter via torch.save

### LINKED ISSUE #72129: torch.save does not save python attributes attached to Tensors or Parameters
### 🐛 Describe the bug

If one attaches python attributes to a Tensor or Parameter, these attributes are not saved by torch.save. This is not expected based on torch's documentation that ["torch.save() and torch.load() use Python’s pickle by default"](https://pytorch.org/docs/stable/notes/serialization.html#saving-loading-tensors).

```python
import torch
xx = torch.zeros(3, 4)
xx.foo = 'bar'
torch.save(xx, '_xx.pt')
torch.load('_xx.pt').foo
```
yields an error
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/tmp/ipykernel_19139/1249207437.py in <module>
----> 1 torch.load('_xx.pt').foo

AttributeError: 'Tensor' object has no attribute 'foo'
```
whereas one would expect this returns `'bar'`.

### Versions

Collecting environment information...
PyTorch version: 1.10.2
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-5.4.0-1067-azure-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 495.46
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.10.2
[pip3] torchaudio==0.10.2
[pip3] torchvision==0.11.3
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.3.1               h2bc3f7f_2
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.3.0           h06a4308_520
[conda] mkl-service               2.4.0            py39h7f8727e_0
[conda] mkl_fft                   1.3.1            py39hd3c417c_0
[conda] mkl_random                1.2.2            py39h51133e4_0
[conda] numpy                     1.21.2           py39h20f2e39_0
[conda] numpy-base                1.21.2           py39h79a1101_0
[conda] pytorch                   1.10.2          py3.9_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.10.2               py39_cu113    pytorch
[conda] torchvision               0.11.3               py39_cu113    pytorch

cc @ezyang @gchanan @zou3519 @mruberry

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes: https://github.com/pytorch/pytorch/issues/72129

TODO:
* [x] Fix for Parameter

Benchmark
(Measurable diff for small tensors)
```
[-------------- Save and Load --------------]
                    |  After PR  |  Before PR
1 threads: ----------------------------------
      ()            |    111.7   |     106.9 
      (4, 4)        |    114.4   |     109.2 
      (128, 128)    |    135.2   |     128.3 
      (1024, 1024)  |   1431.9   |    1431.3 

Times are in microseconds (us).
```

<details>

<summary> Benchmark Script </summary>

```python
import torch
from torch.testing._internal.common_utils import BytesIOContext
from torch.utils import benchmark
import pickle

shapes = ((), (4, 4), (128, 128), (1024, 1024))

sizes = [1, 64, 1024, 10000]
results = []

def save_load_fn(t):
    with BytesIOContext() as f:
        torch.save(t, f)
        f.seek(0)
        torch.load(f)

for shape in shapes:
    t = torch.randn(shape)
    label = 'Save and Load'
    sub_label = f'{shape}'
    results.append(benchmark.Timer(
        stmt='save_load_fn(t)',
        globals={'t': t, 'save_load_fn':save_load_fn},
        label=label,
        sub_label=sub_label,
        description='Before PR',
    ).blocked_autorange(min_run_time=2))

compare = benchmark.Compare(results)
compare.print()

with open('before_pr.pkl', 'wb') as f:
    pickle.dump(results, f)

# with open('after_pr.pkl', 'rb') as f:
#     after_pr = pickle.load(f)

# with open('before_pr.pkl', 'rb') as f:
#     before_pr = pickle.load(f)

# compare = benchmark.Compare(after_pr + before_pr)
# compare.print()
```

</details>

NOTE : **BC-Breaking** : After this PR, all tensors (also regular tensors) will be serialised using `_rebuild_from_type_v2`.

cc @ezyang @gchanan

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Thanks for this fix! This code is way too complex :( 
Only small nits

**Comment by jeanschmidt** [comment]
this commit breaks internal tests/builds - https://pastebin.com/fU9T0NMd https://pastebin.com/1fwiE9ZU

**Comment by albanD** [comment]
@jeanschmidt the first one seems to be a network error?

**Comment by albanD** [comment]
@jeanschmidt not sure what to do with the recond one. Is there any instruction to repro this?

**Comment by jeanschmidt** [comment]
Yeah, the first one seems to be an internal issue in the build system, the 2nd one seems to be the process dying while running tests (bad pointer?).

I believe you don't have access to the code to reproduce it right?

**Comment by albanD** [comment]
> I believe you don't have access to the code to reproduce it right?

Well I am at Meta, but I'm still not sure how to repro that from the pastes? Having a repro that I can run in OSS would help a lot as the PR's author could look into it directly.

**Comment by albanD** [comment]
@kshitij12345 the one line error message is `what():  Unpickler found unknown type torch._tensor._rebuild_from_type_v2` FYI
I'm not sure how that is possible and how this only happens for that one test that seems to be trying to serialize layer norm layer.
Any ideas?

**Comment by kshitij12345** [comment]
@albanD Thanks for pointing that. I don't have a clear idea of what's happening.
The error looks to be from trying to deserialise a pickle on C++ side. Will try to see if I can reproduce something similar locally and take it from there.

Point of failure in C++:
https://github.com/pytorch/pytorch/blob/74c6a754a570c092837dc2b70133c1fb0acce0bf/torch/csrc/jit/serialization/unpickler.cpp#L688-L693

**Comment by albanD** [comment]
@kshitij12345 any update on this by any chance?

**Comment by kshitij12345** [comment]
I'll try to check it by tomorrow and see if I can repro it.

**Comment by kshitij12345** [comment]
Following are the steps to reproduce the above failure

Step 1 : Pickle a tensor (with PR's patch)
```python
import torch
t = torch.randn(3, 3)
torch.save(t, 'test.pt')
```

Step 2: Load the pickled data in C++
```c++
#include <torch/torch.h>
#include <iostream>
#include <vector>

int main() {
    // get pickled data
    std::ifstream file("test.pt");
    std::ostringstream ss;
    ss << file.rdbuf();
    const std::string& s = ss.str();
    std::vector<char> data(s.begin(), s.end());

    // interpret pickled data
    auto a = torch::pickle_load(data);
    std::cout << "FIN\n";
}
```

The reason it fails is because C++ unpickling code has if-else statements to check which function pickled the Tensor and rebuild a Tensor from it. However, there is no corresponding rebuilding case for the new path from `_rebuild_from_type_v2`.
For solving this, if we are to write in C++ unpickler how to rebuild when we used `_rebuild_from_type_v2` should we silently ignore Python State tied with the Tensor? (Is it useful while dealing in C++?) (Does C++ Tensor store a reference Python state dict?)
Ref: https://github.com/pytorch/pytorch/blob/487c0e1181f4e6275a50ed675b50bf1a420beb0d/torch/csrc/jit/serialization/unpickler.cpp#L609-L616

cc: @albanD

**Comment by albanD** [comment]
Thanks for looking into this!
Does that mean that we will get the exact same error if a subclass of Tensor is saved and you try to load it in c++

We do want to update this indeed to be able to properly load the new format for our Tensors.
I think another special case for `_rebuild_from_type_v2` and there, warn if anything is in the state or the type is not a plain Tensor and then call the regular `rebuildTensor` to build a plain old Tensor for this. WDYT?

**Comment by kshitij12345** [comment]
> Does that mean that we will get the exact same error if a subclass of Tensor is saved and you try to load it in c++

Yes. (I haven't tried but I am sure that it will fail).

> We do want to update this indeed to be able to properly load the new format for our Tensors.
I think another special case for _rebuild_from_type_v2 and there, warn if anything is in the state or the type is not a plain Tensor and then call the regular rebuildTensor to build a plain old Tensor for this. WDYT?

Raising a warning makes sense! Will send a patch soon. Thanks!

**Comment by albanD** [comment]
Thanks!

**Comment by albanD** [comment]
@kshitij12345 this is not stale right?

File: torch/_tensor.py [hunk 1]
```diff
@@ -55,9 +55,6 @@ def _rebuild_from_type(func, type, args, dict):
 
 
 def _rebuild_from_type_v2(func, new_type, args, state):
-    if new_type is Tensor:
```
Comment (albanD) [inline torch/_tensor.py:58]: This is BC-breaking right?
Not that we don't want this, but we should document it as such in the release notes.

File: torch/_weights_only_unpickler.py [hunk 2]
```diff
@@ -103,6 +103,11 @@ def _get_allowed_globals():
         torch._utils._rebuild_sparse_csr_tensor,
     ]:
         rc[f"torch._utils.{f.__name__}"] = f
+
+    # Default rebuild function
+    # Handles Tensor Subclasses, Tensor's with attributes.
+    # NOTE: It calls into above rebuild functions for regular Tensor types.
+    rc["torch._tensor._rebuild_from_type_v2"] = torch._tensor._rebuild_from_type_v2
```
Comment (albanD) [inline torch/_weights_only_unpickler.py:110]: What is that file/function?

File: torch/_utils.py [hunk 3]
```diff
@@ -317,6 +318,62 @@ def _rebuild_parameter(data, requires_grad, backward_hooks):
     return param
 
 
+def _rebuild_parameter_v2(data, requires_grad, backward_hooks, state):
```
Comment (albanD) [inline torch/_utils.py:321]: We should add a comment to `_rebuild_parameter` to say that it is now unused and only for BC.

File: torch/csrc/jit/serialization/unpickler.cpp [hunk 4]
```diff
@@ -532,6 +532,21 @@ PickleOpCode Unpickler::readInstruction() {
...
+      // | Dict      | -> (stack_size - 3)
+      // | Key       | -> (stack_size - 2)
+      // | Value     | -> (stack_size - 1)
+      auto stack_size = stack_.size();
+      auto dict_pos = stack_size - 3;
+      auto key_pos = stack_size - 2;
+      auto val_pos = stack_size - 1;
+      auto dict = stack_.at(dict_pos).toGenericDict();
+      dict.insert_or_assign(stack_.at(key_pos), stack_.at(val_pos));
+      stack_.erase(stack_.begin() + (key_pos), stack_.end());
```
Comment (albanD) [inline torch/csrc/jit/serialization/unpickler.cpp:548]: The top of the stack after this points to the Dict. That is expected right?

File: torch/csrc/jit/serialization/unpickler.cpp [hunk 5]
```diff
@@ -546,6 +561,21 @@ PickleOpCode Unpickler::readInstruction() {
 void Unpickler::readGlobal(
     const std::string& module_name,
     const std::string& class_name) {
+  if (this->skip_next_read_global) {
+    this->skip_next_read_global--;
+    if (this->skip_next_read_global == 1) {
```
Comment (albanD) [inline torch/csrc/jit/serialization/unpickler.cpp:566]: nit: empty if is weird.
Maybe just have a comment in there :p

File: torch/csrc/jit/serialization/unpickler.cpp [hunk 6]
```diff
@@ -546,6 +561,21 @@ PickleOpCode Unpickler::readInstruction() {
...
+    } else if (this->skip_next_read_global == 0) {
+      // Corresponds to the type of `Tensor` being unpickled
+      if (module_name != "torch" || class_name != "Tensor") {
+        TORCH_WARN(
+            "Trying to load a Subclassed Tensor, it will converted to at::Tensor in C++");
```
Comment (albanD) [inline torch/csrc/jit/serialization/unpickler.cpp:571]: will be converted

File: torch/csrc/jit/serialization/unpickler.cpp [hunk 7]
```diff
@@ -834,6 +870,31 @@ void Unpickler::rebuildTensor(bool quantized) {
   });
 }
 
+void Unpickler::rebuildTensorFromTypeV2() {
+  this->skip_next_read_global = 2;
```
Comment (albanD) [inline torch/csrc/jit/serialization/unpickler.cpp:874]: Can you add a comment explaining why 2 and how this works?

File: torch/csrc/jit/serialization/unpickler.cpp [hunk 8]
```diff
@@ -834,6 +870,31 @@ void Unpickler::rebuildTensor(bool quantized) {
...
+    //  (function to rebuild base tensor, type of tensor,
+    //   arguments to construct base tensor, Python State (as dict))
+    auto args = pop(stack_).toTuple();
+    size_t tup_idx = 0;
+    const auto args_elems = args->elements();
+    auto base_tensor_args = args_elems.at(tup_idx + 2).toTuple();
+    auto py_state = args_elems.at(tup_idx + 3).toGenericDict();
+    if (py_state.size() > 0) {
+      TORCH_WARN(
+          "Loading Tensor with Python Attribute will be silently ignored");
```
Comment (albanD) [inline torch/csrc/jit/serialization/unpickler.cpp:887]: nit: it is a bit unclear that the attributes will be ignored, not the Tensor.

**Review by albanD — COMMENTED** [review]
SGTM
Just small details.

File: torch/_weights_only_unpickler.py (hunk 2 above)
Comment (kshitij12345) [reply torch/_weights_only_unpickler.py:110]: https://github.com/pytorch/pytorch/blob/20fc474410f49a8b9707297ce778425541a9a131/torch/_tensor.py#L57

Didn't understand the exact question/comment.

File: torch/csrc/jit/serialization/unpickler.cpp (hunk 4 above)
Comment (kshitij12345) [reply torch/csrc/jit/serialization/unpickler.cpp:548]: Yes. 

(Also, SETITEMS does the same thing : https://github.com/pytorch/pytorch/blob/20fc474410f49a8b9707297ce778425541a9a131/torch/csrc/jit/serialization/unpickler.cpp#L433)

File: torch/_weights_only_unpickler.py (hunk 2 above)
Comment (albanD) [reply torch/_weights_only_unpickler.py:110]: Just asking what is this file doing? :p I've never seen it before.

Comment (kshitij12345) [reply torch/_weights_only_unpickler.py:110]: Ah. This was recently added to allow only unpickling of specific types (to partially fix unsafe load using pickle)

See https://github.com/pytorch/pytorch/pull/86812

Comment (albanD) [reply torch/_weights_only_unpickler.py:110]: Ho ok, thanks!

**Comment by kshitij12345** [comment]
@albanD This should be ready for another round. CI failure looks unrelated. PTAL :)

**Comment by kshitij12345** [comment]
@albanD Gentle ping :)

**Review by albanD — APPROVED** [review]
Sounds good!
Sorry for the late review!

**Comment by kurtamohler** [comment]
Does `Pickler::pushLiteralTensor` in `torch/csrc/jit/serialization/pickler.cpp` need to be updated to save the attrs too, or is that already being handled somewhere?

Other than that question, I think this looks good

**Comment by kshitij12345** [comment]
> Does Pickler::pushLiteralTensor in torch/csrc/jit/serialization/pickler.cpp need to be updated to save the attrs too, or is that already being handled somewhere?

I don't think pickler needs to be updated as we are trying to save Python attributes in the pickle. IIUC, C++ Tensor won't have those attributes (or be subclassed Tensor) so I don't think it needs any update on the same.

Is my understanding correct? @kurtamohler 

Thanks!

**Comment by kurtamohler** [comment]
Yeah that makes sense, thanks for checking!

**Review by kurtamohler — APPROVED** [review]

**Comment by kshitij12345** [comment]
Dynamo Failure looks irrelevant

Ref: https://github.com/pytorch/pytorch/actions/runs/3383169877/jobs/5618941610#step:10:73118

```
RuntimeError: test_indexing failed! Received signal: SIGKILL
```

**Comment by mehtanirav** [comment]
Unfortunately, this PR is breaking model publishing internally due BC breaking changes. We need to revert the PR. 

@singlaiiit @jmswen can provide more guidance on how to split the PR, and land it in stages.

@pytorchbot  revert -m "Internal publishing is broken" -c ghfirst

**Comment by mehtanirav** [comment]
cc @albanD on the revert.

**Comment by kshitij12345** [comment]
> @singlaiiit @jmswen can provide more guidance on how to split the PR, and land it in stages.

@singlaiiit @jmswen @albanD  it would be great if you can help me with the same. Thanks!

**Comment by kshitij12345** [comment]
So IIUC, the issue is only for `_rebuild_parameter_v2`. And the fix would be to introduce/define `_rebuild_parameter_v2` in this PR and after 2 weeks we actually start using it via `__reduce_ex__`? Thanks!

File: torch/_utils.py [hunk 9]
```diff
@@ -335,6 +336,65 @@ def _rebuild_parameter(data, requires_grad, backward_hooks):
     return param
 
 
+# Unused for now, but should be used very soon!
+def _rebuild_parameter_v2(data, requires_grad, backward_hooks, state=None):
+    if not state:
```
Comment (kshitij12345) [inline torch/_utils.py:341]: `state` could be `None` for older models.

File: test/test_serialization.py [hunk 10]
```diff
@@ -896,6 +896,26 @@ def test_meta_serialization(self, weights_only):
...
+                f.seek(0)
+                loaded_t = torch.load(f)
+
+            self.assertEqual(t, loaded_t)
+            self.assertEqual(t.foo, loaded_t.foo)
+            self.assertEqual(t.pi, loaded_t.pi)
+
+        t = torch.zeros(3, 3)
+        _test_save_load_attr(t)
+        with self.assertRaises(AttributeError):
```
Comment (kshitij12345) [inline test/test_serialization.py:915]: This will fail till the follow-up PR.

**Comment by ezyang** [comment]
If there aren't any Python attributes, it would also be smart to use the old builder function for FC

**Comment by kshitij12345** [comment]
> If there aren't any Python attributes, it would also be smart to use the old builder function for FC

Right!! Now we only dispatch to the new builders only if the state is present. So this should take care of compatibility! Thank you!

**Review by albanD — COMMENTED** [review]
This is not FC if the user saves a Tensor with state though?

**Comment by ezyang** [comment]
Yes, there is still breakage if there are Python attributes. But there's no way around this, except maybe a kwarg on save that lets you bypass saving python attrs

**Comment by albanD** [comment]
> Yes, there is still breakage if there are Python attributes. But there's no way around this, except maybe a kwarg on save that lets you bypass saving python attrs

No we will warn but not break on the C++ side.
The problem here is that even in python this will break today. We do want the FC window.

**Review by albanD — APPROVED** [review]
SGTM

**Comment by kshitij12345** [comment]
@albanD would you be importing it for internal tests? Or is it ok to merge it once CI is green?

**Comment by albanD** [comment]
This can't be tested internally as it breaks when publishing. So let's go with this as-is!

**Pull request closed without merging**